File: MemeGenerator.java
// MemeGenerator.java
// Meme Generator (GIF) на Java

import java.awt.*;
import java.awt.image.*;
import java.io.*;
import javax.imageio.*;
import java.net.*;
import com.madgag.gif.AnimatedGifEncoder;

public class MemeGenerator {
    public static void main(String[] args) throws Exception {
        String imagePath = null;
        String topText = null;
        String bottomText = null;
        String output = "meme.gif";
        int fontSize = 40;
        String colorHex = "#FFFFFF";
        boolean animate = false;
        int frames = 10;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("-i") || args[i].equals("--image")) {
                imagePath = args[++i];
            } else if (args[i].equals("-t") || args[i].equals("--top")) {
                topText = args[++i];
            } else if (args[i].equals("-b") || args[i].equals("--bottom")) {
                bottomText = args[++i];
            } else if (args[i].equals("-o") || args[i].equals("--output")) {
                output = args[++i];
            } else if (args[i].equals("--font-size")) {
                fontSize = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--color")) {
                colorHex = args[++i];
            } else if (args[i].equals("--animate")) {
                animate = true;
            } else if (args[i].equals("--frames")) {
                frames = Integer.parseInt(args[++i]);
            } else if (args[i].equals("-h") || args[i].equals("--help")) {
                System.out.println("Использование: java MemeGenerator [опции]\n" +
                        "  -i, --image <файл>      Входное изображение\n" +
                        "  -t, --top <текст>       Верхний текст\n" +
                        "  -b, --bottom <текст>    Нижний текст\n" +
                        "  -o, --output <файл>     Выходной GIF\n" +
                        "  --font-size <размер>    Размер шрифта\n" +
                        "  --color <HEX>           Цвет текста\n" +
                        "  --animate               Анимация\n" +
                        "  --frames <число>        Количество кадров");
                return;
            }
        }

        if (imagePath == null || topText == null || bottomText == null) {
            System.err.println("Ошибка: требуются -i, -t, -b");
            System.exit(1);
        }

        // Читаем изображение
        BufferedImage img = ImageIO.read(new File(imagePath));
        if (img == null) {
            System.err.println("Не удалось прочитать изображение");
            System.exit(1);
        }

        // Парсим цвет
        Color textColor = Color.WHITE;
        if (!colorHex.equals("#FFFFFF")) {
            try {
                textColor = Color.decode(colorHex);
            } catch (Exception e) {}
        }

        int totalFrames = animate ? frames : 1;
        AnimatedGifEncoder encoder = new AnimatedGifEncoder();
        encoder.start(output);
        encoder.setRepeat(0);
        encoder.setDelay(500);

        for (int i = 0; i < totalFrames; i++) {
            BufferedImage frame = createFrame(img, topText, bottomText, fontSize, textColor, i, totalFrames, animate);
            encoder.addFrame(frame);
        }
        encoder.finish();
        System.out.println("Мем сохранён в " + output);
    }

    private static BufferedImage createFrame(BufferedImage img, String topText, String bottomText,
                                             int fontSize, Color color, int frameIdx, int totalFrames, boolean animate) {
        int width = img.getWidth();
        int height = img.getHeight();
        BufferedImage frame = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = frame.createGraphics();
        g.drawImage(img, 0, 0, null);

        // Шрифт
        Font font = new Font("Arial", Font.BOLD, fontSize);
        g.setFont(font);
        g.setColor(color);

        // Анимация
        String topDisplay = topText;
        String bottomDisplay = bottomText;
        if (animate && totalFrames > 1) {
            double progress = (double) frameIdx / (totalFrames - 1);
            int topLen = (int) Math.max(1, topText.length() * progress);
            int bottomLen = (int) Math.max(1, bottomText.length() * progress);
            topDisplay = topText.substring(0, Math.min(topLen, topText.length()));
            bottomDisplay = bottomText.substring(0, Math.min(bottomLen, bottomText.length()));
        }

        // Верхний текст
        if (!topDisplay.isEmpty()) {
            FontMetrics fm = g.getFontMetrics();
            int textWidth = fm.stringWidth(topDisplay);
            int x = (width - textWidth) / 2;
            int y = 10 + fm.getAscent();
            g.drawString(topDisplay, x, y);
        }

        // Нижний текст
        if (!bottomDisplay.isEmpty()) {
            FontMetrics fm = g.getFontMetrics();
            int textWidth = fm.stringWidth(bottomDisplay);
            int x = (width - textWidth) / 2;
            int y = height - 10 - fm.getDescent();
            g.drawString(bottomDisplay, x, y);
        }

        g.dispose();
        return frame;
    }
}

File: MemeGenerator.cs
// MemeGenerator.cs
// Meme Generator (GIF) на C#

using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using AnimatedGif;

class MemeGenerator
{
    static void Main(string[] args)
    {
        string imagePath = null;
        string topText = null;
        string bottomText = null;
        string output = "meme.gif";
        int fontSize = 40;
        Color color = Color.White;
        bool animate = false;
        int frames = 10;

        for (int i = 0; i < args.Length; i++)
        {
            if (args[i] == "-i" || args[i] == "--image") imagePath = args[++i];
            else if (args[i] == "-t" || args[i] == "--top") topText = args[++i];
            else if (args[i] == "-b" || args[i] == "--bottom") bottomText = args[++i];
            else if (args[i] == "-o" || args[i] == "--output") output = args[++i];
            else if (args[i] == "--font-size") fontSize = int.Parse(args[++i]);
            else if (args[i] == "--color")
            {
                string hex = args[++i];
                color = ColorTranslator.FromHtml(hex);
            }
            else if (args[i] == "--animate") animate = true;
            else if (args[i] == "--frames") frames = int.Parse(args[++i]);
            else if (args[i] == "-h" || args[i] == "--help")
            {
                Console.WriteLine(@"Использование: MemeGenerator [опции]
  -i, --image <файл>      Входное изображение
  -t, --top <текст>       Верхний текст
  -b, --bottom <текст>    Нижний текст
  -o, --output <файл>     Выходной GIF
  --font-size <размер>    Размер шрифта
  --color <HEX>           Цвет текста
  --animate               Анимация
  --frames <число>        Количество кадров");
                return;
            }
        }

        if (string.IsNullOrEmpty(imagePath) || string.IsNullOrEmpty(topText) || string.IsNullOrEmpty(bottomText))
        {
            Console.WriteLine("Ошибка: требуются -i, -t, -b");
            Environment.Exit(1);
        }

        using (Image img = Image.FromFile(imagePath))
        {
            int totalFrames = animate ? frames : 1;
            using (var gif = AnimatedGif.AnimatedGif.Create(output, 50))
            {
                for (int i = 0; i < totalFrames; i++)
                {
                    using (Bitmap frame = CreateFrame(img, topText, bottomText, fontSize, color, i, totalFrames, animate))
                    {
                        gif.AddFrame(frame, delay: 500);
                    }
                }
            }
        }
        Console.WriteLine($"Мем сохранён в {output}");
    }

    static Bitmap CreateFrame(Image img, string topText, string bottomText, int fontSize, Color color,
                              int frameIdx, int totalFrames, bool animate)
    {
        Bitmap bitmap = new Bitmap(img.Width, img.Height);
        using (Graphics g = Graphics.FromImage(bitmap))
        {
            g.DrawImage(img, 0, 0, img.Width, img.Height);
            using (Font font = new Font("Arial", fontSize, FontStyle.Bold))
            {
                string topDisplay = topText;
                string bottomDisplay = bottomText;
                if (animate && totalFrames > 1)
                {
                    double progress = (double)frameIdx / (totalFrames - 1);
                    int topLen = (int)Math.Max(1, topText.Length * progress);
                    int bottomLen = (int)Math.Max(1, bottomText.Length * progress);
                    topDisplay = topText.Substring(0, Math.Min(topLen, topText.Length));
                    bottomDisplay = bottomText.Substring(0, Math.Min(bottomLen, bottomText.Length));
                }

                // Верхний текст
                if (!string.IsNullOrEmpty(topDisplay))
                {
                    SizeF size = g.MeasureString(topDisplay, font);
                    float x = (img.Width - size.Width) / 2;
                    float y = 10;
                    g.DrawString(topDisplay, font, new SolidBrush(color), x, y);
                }

                // Нижний текст
                if (!string.IsNullOrEmpty(bottomDisplay))
                {
                    SizeF size = g.MeasureString(bottomDisplay, font);
                    float x = (img.Width - size.Width) / 2;
                    float y = img.Height - size.Height - 10;
                    g.DrawString(bottomDisplay, font, new SolidBrush(color), x, y);
                }
            }
        }
        return bitmap;
    }
}
